### ISSUE TITLE
Cannot deploy to multiple servers with same hostname

### ISSUE THREAD

**Issue by adamgraves-choices** [capistrano/capistrano#784]
With the following configuration:

server 'web.example.com', user: 'deploy', roles: %w{web app db}, port: 2201
server 'web.example.com', user: 'deploy', roles: %w{web app db}, port: 2202

Capistrano only runs on a single machine. Is it possible to do what I want with the current setup?

Cheers

**Comment by Envek** [comment]
#1368 successfully ressurects this issue.

What should I do in the case of
```ruby
server '192.168.1.123', user: 'user', port: '12345', roles: %w[ web ], no_release: true
server '192.168.1.123', user: 'user', port: '12346', roles: %w[ app db web ]
server '192.168.1.123', user: 'user', port: '12347', roles: %w[ app web ]
```

`cap stage deploy:check` does absolutely nothing (as I have understand all servers goes merged into one with `:no_release` option as it is implemented in #1368).

Naming these servers with different names in `~/.ssh/config` helps.

Is there any way to avoid merging these servers into single one without `~/.ssh/config` edits?

**Comment by kirs** [comment]
@leehambley what if we could add `merge: false` option to the `server` DSL to solve the issue described by @Envek?

**Comment by leehambley** [comment]
I'm really not in favour of that, since I honestly think that the project-local SSH config is enough, since it goes along with the project in the SCM repository… but it seems whatever we do with the server merging, it's wrong in some cases for some people, but when we take a hash of *server properties* and we then also take some behavioural options in the same hash, with keys with special meanings, I suspect we might make the API worse.

**Comment by mike-stewart** [comment]
Aside from downgrading to 3.3.5, is there a workaround for if I have two servers that share the same host, but have different ports? I.e. if port forwarding is being used to share one domain between two servers. @townsen

**Comment by kirs** [comment]
Mike, as a workaround you can set up two different hosts with the same
hostname in the .ssh/config file.

**Comment by leehambley** [comment]
It's not a workaround, and you can see it documented here, https://github.com/capistrano/sshkit#tunneling-and-other-related-ssh-themes you can also commit an ssh config to your repository and we'll honour it.

**Comment by leehambley** [comment]
also @mike-stewart I'll extend you the same offer as in the other open issue about host identity, provide test cases for your intended use-cases, and we'll work with you to get the tests up to scratch and get your desired behaviour in, if we have unclear, untested cases, it's a perfect opportunity to plant your flag and make an impression on how you would *like* it to work, our test suite is very approachable, especially the unit tests.

**Comment by dreyks** [comment]
Here's the failing spec
```ruby
# spec/integration/dsl_spec_failing.rb

require 'spec_helper'

describe Capistrano::DSL do

  let(:dsl) { Class.new.extend Capistrano::DSL }

  before do
    Capistrano::Configuration.reset!
  end

  describe 'when defining two servers on same host' do
    before do
      dsl.server 'user_a@example.com:1234', roles: %w(app)
      dsl.server 'user_b@example.com:5678', roles: %w(web)
    end

    it 'can fetch the first server' do
       hosts = dsl.roles(:app).map { |server| "#{server.user}@#{server.hostname}:#{server.port}" }
       expect(hosts[0]).to eq 'user_a@example.com:1234'
    end

    it 'can fetch the second server' do
      hosts = dsl.roles(:web).map { |server| "#{server.user}@#{server.hostname}:#{server.port}" }
      expect(hosts[0]).to eq 'user_b@example.com:5678'
    end
  end
end
```

**Comment by pwaterz** [comment]
I'm trying to do the same thing using docker containers locally and am having the same issue. While the .ssh/config option works, I think working out this issue is worth it. The reason being to reduce the required steps for setting up an environment. With docker you basically can get your setup down to one command, except for a few configuration options. This would be another additional option that would have be documented and explained to developers.

**Comment by mattbrictson** [comment]
@pwaterz Since this issue is closed, can we move this discussion to #1427? I believe #1427 is the same problem.

**Comment by leehambley** [comment]
SSHKit and hence Capistrano will pick up a .ssh/config file checked into
your project for cases such as this.

(Sent from my Nexus 6, please forgive typos!)
